### ISSUE TITLE
define dataset variables for Dataset Metadata Check against Define XML

### ISSUE THREAD

**Issue by SFJohnson24** [cdisc-org/cdisc-rules-engine#791]
currently for rules that are Dataset Metadata Check against Define XML, the ContentsDefineDatasetBuilder does not report define_dataset_variables.  Having this produced as one of the pieces of metadata would be useful for rule development

**Comment by SFJohnson24** [comment]
see also: https://github.com/cdisc-org/cdisc-rules-engine/issues/784

**Comment by mhungria** [comment]
I look forward to this enhancement.

**Comment by gerrycampion** [comment]
I think we should instead use: `define_dataset_variable_order`
This will eliminate the need for #784 `get_column_order_from_define` operation

**Comment by mhungria** [comment]
I think there may be confusion on the functionality I requested with get_column_order_from_define ("Returns list of dataset columns in order" as per the Define-XML document; i.e., list of ItemGroupDef/ItemRef ordered by ItemGroupDef/ItemRef/@OrderNumber. It should be similar to get_column_order_from_dataset ("Returns list of dataset columns in order" as included in the actual dataset).
Note the metadata variable define_variable_order_number (ItemGroupDef.ItemRef.OrderNumber) is only one value, not a list.

**Comment by gerrycampion** [comment]
@mhungria, I think we are on the same page. My idea is that given the following XML:

```xml
<ItemGroupDef OID="IG.SE" Name="SE" Domain="SE">
    <ItemRef ItemOID="IT.SE.STUDYID" OrderNumber="1"/>
    <ItemRef ItemOID="IT.SE.DOMAIN" OrderNumber="2"/>
    <ItemRef ItemOID="IT.SE.USUBJID" OrderNumber="3"/>
    <ItemRef ItemOID="IT.SE.ETCD" OrderNumber="5"/>
    <ItemRef ItemOID="IT.SE.SESEQ" OrderNumber="4"/>
</ItemGroupDef>
```

`define_dataset_variable_order` will return a list with:
- STUDYID
- DOMAIN
- USUBJID
- SESEQ*
- ETCD*

*Note that ETCD and SESEQ are switched because `OrderNumber` attribute exists

given the following XML:
```xml
<ItemGroupDef OID="IG.SE" Name="SE" Domain="SE">
    <ItemRef ItemOID="IT.SE.STUDYID"/>
    <ItemRef ItemOID="IT.SE.DOMAIN"/>
    <ItemRef ItemOID="IT.SE.USUBJID"/>
    <ItemRef ItemOID="IT.SE.ETCD"/>
    <ItemRef ItemOID="IT.SE.SESEQ"/>
</ItemGroupDef>
```
`define_dataset_variable_order` will return a list with:
- STUDYID
- DOMAIN
- USUBJID
- ETCD
- SESEQ

Since there is no `OrderNumber` attribute, it follows the element order.

![Image](https://github.com/user-attachments/assets/64b0c9d1-39dd-448b-8631-23d6a2954adc)

Is there anything this is missing?

**Comment by mhungria** [comment]
The examples for the two use cases where a user wants to retrieve the list of variables in order look ok - thanks!

However, regarding my note "Note the metadata variable define_variable_order_number (ItemGroupDef.ItemRef.OrderNumber) is only one value, not a list.", I would expect to be able to retrieve that value for other use cases that a user needs a variable's order number value as indicated explicitly in the Define-XML document, or implicitly if ItemGroupDef/ItemRef/@OrderNumber is not provided.

**Comment by gerrycampion** [comment]
Can you give an example rule where you need [define_variable_order_number](https://cdisc-org.github.io/conformance-rules-editor/#/metadata_variables?id=define_variable_order_number) and it is not working properly?

**Comment by mhungria** [comment]
Last time I tested that functionality,  I was looking at a define_variable_order_number case with explicit order and I think it was ok. My note is to clarify that we would need both define_variable_order_number metadata variable and get_column_order_from_define operation.
If I understand it correctly, 1) the define_variable_order_number metadata variable is available for rule type Define Item Metadata Check and 2) I assume the get_column_order_from_define operation would be available for rule type Dataset Metadata Check against Define XML. Is that correct?

I was just cross-checking the reference guide and could not find documentation for define_dataset_variables for rule type Dataset Metadata Check against Define XML. Was that not implemented?

**Comment by gerrycampion** [comment]
> 1) the define_variable_order_number metadata variable is available for rule type Define Item Metadata Check 

Yes, it should be because this is __variable-level__

> 2) I assume the get_column_order_from_define operation would be available for rule type Dataset Metadata Check against Define XML. Is that correct?

I think it would be better to use a metadata variable instead of operation, but yes it would be for `Dataset Metadata Check against Define XML` because it would be __dataset-level__

> I was just cross-checking the reference guide and could not find documentation for define_dataset_variables for rule type Dataset Metadata Check against Define XML. Was that not implemented?

It should be: https://cdisc-org.github.io/conformance-rules-editor/#/rule_type?id=dataset-metadata-check-against-define-xml

**Comment by mhungria** [comment]
#1: ok, thanks.
#2: "I think it would be better to use a metadata variable instead of operation", sure. I assume the documentation will clearly indicate that it will return a list of variables in the explicit (or, implicit) order.
#3: the documentation "https://cdisc-org.github.io/conformance-rules-editor/#/rule_type?id=dataset-metadata-check-against-define-xml" only lists the variable but it doesn't include a description of the value returned.

**Comment by gerrycampion** [comment]
Ok, I created a ticket for the missing docs: #1045

**Comment by RamilCDISC** [comment]
@mhungria Could you please provide a rule that uses define_dataset_variable_order as your intended use. It will help me in thorough validation.

**Comment by mhungria** [comment]
@gerrycampion, I'll have to draft a rule that can be used for testing.

**Here is a set of files you can use for testing for _Rule CG0019_. The results reflected are not expected at all.**
If needed, we can TC when convenient for both of us.

[Rule (20).zip](https://github.com/user-attachments/files/24629226/Rule.20.zip)

Note:
In the enclosed yaml file for draft rule CG0019, I commented the intended operator and replaced with the opposite to ensure output variables.

Check:
  all:
    - name: define_dataset_key_sequence
      **#operator: is_not_unique_set
      operator: is_unique_set**
Core:
  Id: CDISC.SDTMIG.CG0019
...
  Output Variables:
    - define_dataset_key_sequence
    - **define_dataset_variable_order**
